### PR TITLE
[DAG] Add SDPatternMatch m_ZExtOrSelf/m_SExtOrSelf/m_AExtOrSelf/m_TruncOrSelf matchers

### LINKED ISSUE #85395: [SelectionDAG] Add m_ZExtOrSelf/m_SExtOrSelf/m_AExtOrSelf/m_TruncOrSelf pattern matchers
Followup to #85065

Similar to what we have in PatternMatch.h - allows us to peek though optional extensions/truncations

CC @mshockwave

### PR DESCRIPTION
**Pull request by marcauberer**
Fixes #85395

### CODE REVIEW

**Comment by marcauberer** [comment]
@RKSimon @mshockwave this is now ready for review.
I am not entirely sure how I would add negative tests, since an expression does always match with e.g. `m_ZExtOrSelf()`, because it is always "self". Maybe take a larger expression from InstCombine, that uses the InstCombine `m_ZExtOrSelf()` and modify it for the DAG unit test? What do you think?

File: llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp [hunk 1]
```diff
@@ -251,6 +251,29 @@ TEST_F(SelectionDAGPatternMatchTest, patternCombinators) {
...
+
+  using namespace SDPatternMatch;
+  EXPECT_TRUE(sd_match(Op32, m_ZExtOrSelf(m_Value())));
+  EXPECT_TRUE(sd_match(ZExt, m_ZExtOrSelf(m_Value())));
+  EXPECT_TRUE(sd_match(Op64, m_SExtOrSelf(m_Value())));
+  EXPECT_TRUE(sd_match(SExt, m_SExtOrSelf(m_Value())));
+  EXPECT_TRUE(sd_match(Op32, m_AExtOrSelf(m_Value())));
+  EXPECT_TRUE(sd_match(AExt, m_AExtOrSelf(m_Value())));
+  EXPECT_TRUE(sd_match(Op64, m_TruncOrSelf(m_Value())));
+  EXPECT_TRUE(sd_match(Trunc, m_TruncOrSelf(m_Value())));
```
Comment (RKSimon) [inline llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp:274]: For these tests you will need to pass in an argument to m_Value() and then test that it's set to the right value

Comment (marcauberer) [reply llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp]: Done.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 2]
```diff
@@ -737,6 +737,39 @@ template <typename ValTy>
...
+                                        std::forward<Opnd>(Op));
+}
+
+/// Match a trunc or identity
+/// Allows to peek through optional truncations
+template <typename Opnd>
+inline Or<UnaryOpc_match<Opnd>, Opnd> m_TruncOrSelf(Opnd &&Op) {
+  return Or<UnaryOpc_match<Opnd>, Opnd>(m_Trunc(std::forward<Opnd>(Op)),
+                                        std::forward<Opnd>(Op));
+}
```
Comment (RKSimon) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:771]: (pedantic) but would it be possible to move these next to the other m_*ext/m_Trunc variants?

Comment (marcauberer) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done.

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by marcauberer** [comment]
@RKSimon can you merge? I do not have commit permissions.

**Pull request merged**